Delay smt-solving until it is absolutely necessary in EVM

gustavo-grieco opened the pull request:
This PR contains:

* A lazy evaluation mode for `jumpi` constraints enabled by default (so we can see the effects in the tests). This mode cannot be used if the code contains a loop. There is no automatic detection of that, so the user should be careful when using this mode. It is unclear how to handle that in the middle of a manticore run.
* Various fixed for the evm tests to make sure the states are always sound.
* `test_jmpdest_check` test was disabled.
* Some tests using plugins require lazy mode to be disabled.

The last two bullets should be discussed before merging this PR.

gustavo-grieco committed:
delay smt-solving until it is absolutely necessary

gustavo-grieco committed:
linted

gustavo-grieco committed:
force finalization in the VMTests

gustavo-grieco committed:
fixed unsoundness tests

gustavo-grieco committed:
fix

gustavo-grieco committed:
merge

gustavo-grieco committed:
lint

gustavo-grieco committed:
fixes

gustavo-grieco committed:
fixes

gustavo-grieco committed:
disable lazy evaluation for verifier

gustavo-grieco committed:
fixed more tests

gustavo-grieco committed:
more fixes

gustavo-grieco committed:
fix

ehennenfent commented on manticore/core/state.py:
```diff
@@ -360,7 +360,7 @@ def new_symbolic_value(self, nbits, label=None, taint=frozenset()):
         return expr
 
-    def concretize(self, symbolic, policy, maxcount=7):
+    def concretize(self, symbolic, policy, maxcount=7, additional_symbolics=None):
```
I'm not thrilled with introducing an additional argument to a user-facing function if it's only used in one other place in Manticore. I'd suggest either:
* Removing the `additional_symbolics` argument from `concretize` and instead calling `concretize` twice in ethereum/manticore.py and zipping the results (since it results in the same number of solver calls regardless)

_Or_

* Adding type hints and documentation items describing each argument, and making `additional_symbolics` keyword-only.

ggrieco-tob replied:
Where I should add the type hint exactly for `additional_symbolics`?

ehennenfent replied:
Right on this line :). So you'd get something that looks like `def concretize(self, symbolic, policy, maxcount=7, *, additional_symbolics: typing.Optional[typing.List[Expression]]=None):`

ehennenfent commented on manticore/core/state.py:
```diff
@@ -412,9 +412,19 @@ def concretize(self, symbolic, policy, maxcount=7):
-                self._constraints, symbolic, maxcnt=maxcount, silent=True
-            )
+            if additional_symbolics is not None:
+                print(symbolic, additional_symbolics)
```
If you do end up leaving `additional_symbolics`, this should be swapped out for `logger.debug` or `logger.info`

ggrieco-tob replied:
Fixed.

ehennenfent commented on manticore/platforms/evm.py:
```diff
@@ -2539,6 +2548,12 @@ def constraints(self, constraints):
+    def last_ins_was_true_jumpi(self) -> Optional[bool]:
+        return self._last_ins_was_true_jumpi
+
+    def set_last_ins_was_true_jumpi(self, jump_cond: Optional[bool]):
```
Let's convert this getter/setter pairing to a property as well.

ggrieco-tob replied:
Fixed

ehennenfent commented on manticore/platforms/evm.py:
```diff
@@ -863,6 +871,18 @@ def constraints(self, constraints):
     def gas(self):
         return Operators.EXTRACT(self._gas, 0, 256)
 
+    def jumpi_false_branch(self) -> Optional[int]:
```
Python has a nifty replacement for getter & setter methods called [Properties](https://docs.python.org/3/library/functions.html#property). Since it looks like these don't explicitly need to be callable, we should convert them to properties.

ggrieco-tob replied:
Fixed.

ehennenfent commented on manticore/ethereum/manticore.py:
```diff
@@ -1953,3 +1969,100 @@ def worker_finalize(q):
+        """
+        self._lazy_evaluation = flag
+
+    def get_lazy_evaluation(self) -> bool:
```
As before, a property would be perfect here. As for `set_lazy_evaluation`, you could convert that to a property as well, but I think this is intended to be user-facing, and most users will trust `manticore.set_lazy_evaluation(False)` more than `manticore.lazy_evaluation = False`. That said, to make it even more explicit, you could consider splitting this up into `enable_lazy_evaluation` and `disable_lazy_evaluation` methods.

ggrieco-tob replied:
Fixed, now we have `enable/disable_lazy_evaluation`.

gustavo-grieco committed:
fixes

gustavo-grieco committed:
fixes

gustavo-grieco committed:
more fixes

gustavo-grieco committed:
fix

Commit:
Tweak typing

Commit:
Disable lazy evaluation by default

Commit:
Merge branch 'master' into dev-lazy-mode

gustavo-grieco committed:
Merge branch 'master' of ssh://github.com/trailofbits/manticore into dev-lazy-mode

gustavo-grieco committed:
Merge branch 'dev-lazy-mode' of ssh://github.com/trailofbits/manticore into dev-lazy-mode

gustavo-grieco committed:
fix

gustavo-grieco committed:
test fix

Commit:
Remove disable_lazy_evaluation calls

It's already disabled and we want to look at these tests again if we enable it by default in the future.

Commit:
Merge branch 'master' into dev-lazy-mode

Commit:
Re-enable jumpdest check

ehennenfent commented:
@ggrieco-tob I can't replicate the failing tests locally. I believe you had the same problem as well. I'd say let's hold this one back from the next release? That will give us more time to figure out what's going on.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/trailofbits/manticore?pullRequest=2425) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/trailofbits/manticore?pullRequest=2425) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: ggrieco-tob<br/>:x: Eric Hennenfent<br/><hr/>**Eric Hennenfent** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/trailofbits/manticore?pullRequest=2425) it.</sub>
